Run bigscape v2

adraismawur opened the pull request:
#216 
This implements running BiG-SCAPE 2.

I opted for keeping the run_bigscape function the same and changing how the arguments are built up to avoid a large amount of duplicated code. Let me know if a separate run_bigscape_v2 function is still desired.

The installation of BiG-SCAPE 2 will likely change in time. I have now fixed it to a reliable development version by commit, mirroring the BiG-SCAPE 1 installation. It's limited to low blob filesizes so should only minimally impact the installation time.

When we package BiG-SCAPE 2 we should be able to easily change this installation.

adraismawur committed:
add BiG-SCAPE 2 to dependencies

adraismawur committed:
add example config for bigscape 2

adraismawur committed:
implement running BiG-SCAPE

adraismawur committed:
fix bigscape2 dependency

adraismawur committed:
copy db file properly

adraismawur committed:
remove cluster arg

adraismawur committed:
run ruff formatter

adraismawur committed:
fix ruff check issues

adraismawur committed:
ensure str for mypy static type checking

adraismawur commented:
Seems like a few unrelated tests aren't passing? Let me know if I broke something.

CunliangGeng commented on bin/install-nplinker-deps:
```diff
@@ -125,6 +125,7 @@ pip install -q -U pip setuptools
 echo "🔥 Start installing BigScape ..."
     [[ -d BiG-SCAPE ]] || git clone https://github.com/medema-group/BiG-SCAPE.git
     cd BiG-SCAPE
+    git reset --hard
```
Why is `git reset --hard` needed here?

adraismawur replied:
I added this during testing of installation and forgot to remove it.

However, this does ensure that when you re-run install-nplinker-deps, all tasks complete successfully.
Otherwise, this script crashes here because you first need to stash/commit/revert changes before you do the other steps.

Would it be practical to just skip all of this entirely if the BiG-SCAPE directory already exists?

CunliangGeng replied:
Interesting, I did not see any changes to the local code repo. Why is it necessary to stash/commit/revert **changes** before you do the other steps? Is there anything I miss?

adraismawur replied:
The modification of permissions here:

```bash
    chmod 754 bigscape.py
    chmod 664 domains_color_file.tsv
    chmod 775 Annotated_MIBiG_reference
```

Is being tracked and results in conflicts on a re-run for me.

CunliangGeng replied:
Aha, indeed! OK, then we can keep the `git reset --hard` to solve the local changes problem.

CunliangGeng commented on bin/install-nplinker-deps:
```diff
@@ -136,6 +137,21 @@ echo "🔥 Start installing BigScape ..."
+    chmod 754 bigscape.py
+    ln -sf $LIB_PATH/BiG-SCAPE-v2/bigscape.py $PY_PATH/bin/bigscape-v2.py
+    ln -sf $LIB_PATH/BiG-SCAPE-v2/config.ini $PY_PATH/bin # new system of configuration needs default config file
+    ln -sf $LIB_PATH/BiG-SCAPE-v2/big_scape $PY_PATH/bin # folder is needed for some files
```
Will the files in `big_scape` change or increase? The `bin` folder is not really a good place for data files.

adraismawur replied:
This should change once we implement packaging and installation (which I am working on today).

I will push another change to get rid of these files.

CunliangGeng commented on src/nplinker/nplinker_default.toml:
```diff
@@ -9,7 +9,12 @@ to_use = true
 version = "3.1"
 
 [bigscape]
+verison = 1
```
Sorry, It might be confusing, but you need to add all the changes in this file to `src/nplinker/data/nplinker.toml`, which is the template file that contains all configs.

This `nplinker_default.toml` file only contains the default values. So here only `version=1` is required to add.

CunliangGeng commented on src/nplinker/nplinker_default.toml:
```diff
@@ -9,7 +9,12 @@ to_use = true
 version = "3.1"
 
 [bigscape]
+verison = 1
```
Since the `version` is a new option, you also need to add an validator for it in the `src/nplinker/config.py` file. 
The validator could be `Validator("bigscape.version", required=True, is_type_of=int),`.

Then it's also necessary to update the `test_config.py` file.

adraismawur replied:
Whoops, unresolved because I didn't add the validators yet.

CunliangGeng commented on src/nplinker/nplinker_default.toml:
```diff
@@ -9,7 +9,12 @@ to_use = true
+
+# for version 2, use the following parameters string:
+# parameters = "--mibig_version 3.1 --include_singletons --gcf_cutoffs 0.30"
+
```
```suggestion
```

Move them to `src/nplinker/data/nplinker.toml` file.

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,46 @@ def run_bigscape(
+    if version == 1:
+        args.extend(["--pfam_dir", PFAM_PATH])
+    elif version == 2:
+        args.extend(["cluster", "--pfam_path", PFAM_PATH + "/Pfam-A.hmm"])
```
```suggestion
        args.extend(["cluster", "--pfam_path", os.path.join(PFAM_PATH,  "Pfam-A.hmm"))
```
The separator `/` does not work on Windows.

CunliangGeng reviewed (changes requested):
The changes look good to me. A few small things need to be updated. But please first solve the conflict to get the latest codebase before addressing my comments.

adraismawur committed:
Merge branch 'dev' of github.com:NPLinker/nplinker into run-bigscape-v2

adraismawur committed:
Move configuration to correct file

adraismawur committed:
use os.path.join instead of string concat

adraismawur committed:
fix merge mistake

CunliangGeng reviewed (changes requested):
There is not unit tests for the `rubbigscape.py`. It'd great to add them with some simple input data.

adraismawur committed:
remove extra bigscape 2 files

adraismawur committed:
add missing library

adraismawur committed:
add validator for bigscape version

adraismawur committed:
add test for bigscape version

adraismawur committed:
fix typo

adraismawur commented:
Sorry this took a while, getting the structure of BiG-SCAPE right to be a normal package was a bit of a hassle.

This should now cover all the issues.

CunliangGeng commented:
Thanks, the new commits look good to me.

The unit tests are still missing for testing the `run_bigscape` function, could you add that?

adraismawur committed:
Merge branch 'dev' of github.com:NPLinker/nplinker into run-bigscape-v2

adraismawur committed:
add simple run testing

adraismawur committed:
add test to check for nonextent input path

adraismawur commented:
I've added some very rudimentary tests, they add --help as an argument so that the process returns successfully quickly.

I also tried adding some where the actual run is tested by inputting some files and checking for the existence of the expected output files, but I am seeing that it takes a relatively long time for v1 to finish, and for some reason v2 doesn't finish at all.

If wanted, I can still try adding this, but in theory the important code is covered with only these fast tests.

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,46 @@ def run_bigscape(
+
+    The behavior of this function is slightly different depending on the version of
+    BiG-SCAPE that is set to run using the configuration file.
+    Mostly this means a different set of parameters is used between the two versions.
```
Please also add the docstring for the arguments and return with a Google stype: 
```
Args:
        antismash_path: Path to ...

Returns:
        ...
```

CunliangGeng replied:
It's better to describe the requirements on the structure of the data in the `antismash_path`, then users would know how to prepare it.

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,46 @@ def run_bigscape(
+    Mostly this means a different set of parameters is used between the two versions.
+    """
+    # switch to correct version of BiG-SCAPE
+    if version == 1:
+        bigscape_py_path = "bigscape.py"
+    elif version == 2:
+        bigscape_py_path = "bigscape-v2.py"
```
An ValueError should be raised if the version is neither 1 or 2.

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,46 @@ def run_bigscape(
+        bigscape_py_path = "bigscape-v2.py"
+
     logger.info(
         f'run_bigscape: input="{antismash_path}", output="{output_path}", extra_params={extra_params}"'
```
Also add bigscape version info to the logging info?

CunliangGeng replied:
it is better to move this logger info after the code of validating all input arguments

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,46 @@ def run_bigscape(
         subprocess.run([bigscape_py_path, "-h"], capture_output=True, check=True)
     except Exception as e:
-        raise Exception(f"Failed to find/run bigscape.py (path={bigscape_py_path}, err={e})") from e
+        raise Exception(
```
We should raise a specific exception here, e.g. `FileNotFoundError`

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,46 @@ def run_bigscape(
+        ) from e
 
     if not os.path.exists(antismash_path):
         raise Exception(f'antismash_path "{antismash_path}" does not exist!')
```
Should replace `Exception` with a specific exception, e.g. `FileNotFoundError`

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
The last part (line 73-end) needs to be refactored to return a more clear message about the outcome of the running. The `CalledProcessError` is not helpful for users to understand I guess.

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
We need to test the actual running to make sure it really works. Maybe make a smaller antismash input to speed up the running?

CunliangGeng replied:
We also need to test all possible exceptions

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,36 @@
+from .. import DATA_DIR
+
+
+class TestRunBigscape:
```
Using test class `class TestRunBigscape` is not necessary here. The code would be simpler if using test functions directly.

CunliangGeng reviewed (commented):
Thanks for the update. I had a look at all changes and related code. 
I leave some comments to improve the (legacy) code to make sure they are aligned with the other refactored code.

adraismawur committed:
add info to docstring

adraismawur committed:
add exception on invalid version number

adraismawur committed:
move log to after validation

adraismawur committed:
add version info to log

adraismawur committed:
use specific exception

adraismawur committed:
rework return codes and exceptions

adraismawur committed:
add wrong version test

adraismawur committed:
add invalid path test for v2

adraismawur committed:
specify exception

adraismawur committed:
fix tests not correctly running

adraismawur committed:
change imports to reflect style in other tests

adraismawur committed:
specify exception type

adraismawur committed:
add minimal test data

adraismawur committed:
add real data tests

adraismawur committed:
remove class

adraismawur committed:
force string for mypy

adraismawur commented:
Finally got to this again! Addressed the remaining issues.

Not sure why mypy is complaining now. I reproduced that mypy error locally when checking runbigscape.py but the github workflow seems to still detect it.

adraismawur commented:
The small dataset included for the BiG-SCAPE tests come from this paper: https://doi.org/10.1016/j.biocontrol.2023.105411

All that should be necessary to check if a BiG-SCAPE run with the default NPLinker settings is if it completes successfully on these four GBKs.

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
+    elif version == 2:
+        bigscape_py_path = "bigscape-v2.py"
+    else:
+        raise ValueError("Unexpected BiG-SCAPE version number specified")
```
```suggestion
        raise ValueError("Invalid BiG-SCAPE version number. Expected: 1 or 2.")
```

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
-    logger.info(
-        f'run_bigscape: input="{antismash_path}", output="{output_path}", extra_params={extra_params}"'
-    )
+    version: int = 1,
```
Since `version` can only be `1` or `2`, we can use [Literal](https://docs.python.org/3.10/library/typing.html#typing.Literal) to annotate the type: `
```suggestion
    version: Literal[1, 2] = 1,
```

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
+        antismash_path: Path to the antismash output directory.
+        output_path: Path to the output directory where BiG-SCAPE will write its results.
+        extra_params: Additional parameters to pass to BiG-SCAPE.
+        version: The version of BiG-SCAPE to run. Can be 1 or 2.
```
```suggestion
        version: The version of BiG-SCAPE to run. Must be 1 or 2.
```

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
+
+    Returns:
+        True if BiG-SCAPE ran successfully, False otherwise.
+    """
```
Please also add `Raises:` (for all exceptions of this function) and `Examples:` (for both bigscape versions and typical extra params) to the docstring. See the [example code](https://nplinker.github.io/nplinker/latest/api/antismash/#nplinker.genomics.antismash.download_and_extract_antismash_data).

CunliangGeng replied:
Adding some `Examples:` to the docstring?

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
     except Exception as e:
-        raise Exception(f"Failed to find/run bigscape.py (path={bigscape_py_path}, err={e})") from e
+        raise FileNotFoundError(
+            f"Failed to find/run bigscape.py (path={bigscape_py_path}, err={e})"
```
```suggestion
            f"Failed to find/run BiG-SCAPE executable program (path={bigscape_py_path}, err={e})"
```

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
+    recursively for GBK files.
+
+    By default, only GBK Files with "cluster" or "region" in the filename are
+    accepted. GBK Files with "final" in the filename are excluded.
```
Is it possible to give an example for the antismash output? like what I give in [AntismashBGCLoader](https://nplinker.github.io/nplinker/latest/api/antismash/)

![image](https://github.com/user-attachments/assets/6d4c3f8f-a342-4844-9e12-10a9a30c1a8d)

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -39,9 +82,13 @@ def run_bigscape(
+
+    # otherwise log details and raise a runtime error
+    logger.error(f"BiG-SCAPE failed with return code {result.returncode}")
+    logger.error(f"output: {str(result.stdout)}")
```
Also log `result.stderr` for debugging?

CunliangGeng commented on nplinker.toml:
Why is this new file necessary?

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,83 @@
+
+def test_run_bigscape_v1(tmp_path):
+    result = bigscape.run_bigscape(
+        antismash_path=DATA_DIR,
```
The antismash path is `DATA_DIR`, which contains many gbk files in sub-folders. Is that what you want?

CunliangGeng replied:
I guess you just want to run `--help`, for that, I prefer using `tmp_path` as the fake antismash path. To make the intention of the test more clear, maybe adding a docstring like "Test `bigscape --help`".

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,83 @@
+
+def test_run_bigscape_small_dataset_v1(tmp_path):
+    result = bigscape.run_bigscape(
+        antismash_path=DATA_DIR / "bigscape/minimal_dataset",
```
```suggestion
        antismash_path=DATA_DIR / "bigscape" / "minimal_dataset",
```

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,22 +15,65 @@ def run_bigscape(
+    if version == 1:
+        args.extend(["--pfam_dir", PFAM_PATH])
+    elif version == 2:
+        args.extend(["cluster", "--pfam_path", os.path.join(PFAM_PATH, "Pfam-A.hmm")])
```
Since some bigscape parameters have been pre-configured, it's needed to update `data/nplinker.toml` to inform users not to set these parameters.

adraismawur replied:
If you mean the "cluster" bit, that's the mode in which BiG-SCAPE v2 runs and separates some functionality that were previously extra parameters.
In particular, query mode is no longer a parameter, but it's own mode run using `bigscape.py query [params]`.

I will add this to the explanation of nplinker.toml

CunliangGeng replied:
There are explanations for version 1 in [nplinker.toml](https://github.com/NPLinker/nplinker/blob/1e23e57a81910ad19964d4c65b0d67be12550362/src/nplinker/data/nplinker.toml#L44), you could take a look. 
If version 2 has subcommands, then we need to make it clear it's using the subcommand `cluster` and the parameter `--pfam_path`. That means we won't allow users to set other subcommand in the config file.

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,83 @@
+
+def test_run_bigscape_small_dataset_v2(tmp_path):
+    result = bigscape.run_bigscape(
+        antismash_path=DATA_DIR / "bigscape/minimal_dataset",
```
```suggestion
        antismash_path=DATA_DIR / "bigscape" / "minimal_dataset",
```

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,83 @@
+    assert "version" in e.value.args[0]
+
+
+def test_input_path_not_exist_v1(tmp_path):
```
Since only version changes for the last two tests, you could use "@pytest.mark.parametrize" to simplify the tests, see `test_gnps_format.py` as an example.

CunliangGeng reviewed (changes requested):
Thanks for the update! Almost there.

I see the unit tests (github action) have been running for two hours ;-) Is it possible to have a lot more smaller antismash input (e.g. fake and mini gbk files) to make sure the running of bigscape complete within 2 mins? If it's not possible, I would prefer discarding these unit tests.

CunliangGeng commented:
I just killed the github actions to stop the unit tests.

adraismawur committed:
Apply suggestions from code review

Co-authored-by: Cunliang Geng <c.geng@esciencecenter.nl>

adraismawur committed:
add exceptions to docstring

adraismawur committed:
add docstring to tests

adraismawur committed:
use tmp path instead of data path

adraismawur committed:
add missing typing

adraismawur committed:
add explanation of cluster mode

adraismawur committed:
parameterize tests

adraismawur commented:
> Thanks for the update! Almost there.
> 
> I see the unit tests (github action) have been running for two hours ;-) Is it possible to have a lot more smaller antismash input (e.g. fake and mini gbk files) to make sure the running of bigscape complete within 2 mins? If it's not possible, I would prefer discarding these unit tests.

Oh no, they should already complete within two minutes... 
I can check if it works with only two gbk files, or maybe I can make them very minimal by trimming the files themselves.
Otherwise I think we unfortunately need to remove those tests.

adraismawur committed:
remove two gbks

adraismawur commented:
Okay it looks like this test just refuses to complete on the worker. I don't understand why; this is a very small sample.

adraismawur committed:
better documentation

adraismawur committed:
skip tests with dataset

adraismawur committed:
do not check output code within run

adraismawur committed:
move log

adraismawur committed:
add test with incorrect parameters for runtime exception

CunliangGeng commented on nplinker.toml:
Remove this `nplinker.toml` file? it's in the root folder and not used by any code.

adraismawur replied:
Whoops, sorry about that. Accidentally included it.

CunliangGeng commented on src/nplinker/data/nplinker.toml:
```diff
@@ -51,6 +54,10 @@ version = "3.1"
 parameters = "--mibig --clans-off --mix --include_singletons --cutoffs 0.30"
+# for version 2, use the following parameters string:
+# parameters = "--mibig_version 3.1 --include_singletons --gcf_cutoffs 0.30"
+# Note that BiG-SCAPE v2 has subcommands. NPLinker requires the "cluster" subcommand to be used.
+# BiG-SCPAPE v2 also runs a --mix analysis by default, and does not need this to be included.
```
```suggestion
# Note that BiG-SCAPE v2 has subcommands. NPLinker requires the "cluster" subcommand and its parameters to be used. 
NPLinker will automatically set the following parameters: `--pfam_path`, `--inputdir` and `--outputdir`. So, do not set them here.
# BiG-SCPAPE v2 also runs a --mix analysis by default, and does not need this to be included.
```

I added the "`--inputdir` and `--outputdir`", maybe it's not right.

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,67 @@
+
+@pytest.mark.parametrize("version", [1, 2])
+def test_run_bigscape_small_dataset(tmp_path, version):
+    pytest.skip("This test is too slow to run in CI")
```
```suggestion
```

CunliangGeng commented on tests/unit/genomics/test_runbigscape.py:
```diff
@@ -0,0 +1,67 @@
+    assert result is True
+
+
+@pytest.mark.parametrize("version", [1, 2])
```
```suggestion
@pytest.mark.skipif(os.getenv('GITHUB_ACTIONS') == 'true', reason="The test is time-consuming on CI")
@pytest.mark.parametrize("version", [1, 2])
```
Need to add `import os`

CunliangGeng reviewed (changes requested):
OK, let's skip the test on CI, but keep running it on local computer, it's really fast.

A few more comments to address. After that, I think we are ready to release a new version of NPlinker.

adraismawur committed:
remove temporary nplinker.toml

adraismawur committed:
add stderr to error log

adraismawur committed:
add import needed for skipping test on CI

adraismawur committed:
Apply suggestions from code review

Co-authored-by: Cunliang Geng <c.geng@esciencecenter.nl>

adraismawur committed:
expand docstring

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,33 +16,98 @@ def run_bigscape(
+    The directory can contain subdirectories, in which case BiG-SCAPE will search
+    recursively for GBK files. E.g.:
+
+    example_folder
+        ├── organism_1
+        │  ├── organism_1.region001.gbk
+        │  ├── organism_1.region002.gbk
+        │  ├── organism_1.region003.gbk
+        │  ├── organism_1.final.gbk          <- skipped!
+        │  └── ...
+        ├── organism_2
+        │  ├── ...
+        └── ...
```
```suggestion
    ```
    example_folder
        ├── organism_1
        │  ├── organism_1.region001.gbk
        │  ├── organism_1.region002.gbk
        │  ├── organism_1.region003.gbk
        │  ├── organism_1.final.gbk          <- skipped!
        │  └── ...
        ├── organism_2
        │  ├── ...
        └── ...
    ```
```

CunliangGeng commented on src/nplinker/genomics/bigscape/runbigscape.py:
```diff
@@ -15,33 +16,98 @@ def run_bigscape(
+        FileNotFoundError: If the antismash_path does not exist or if the BiG-SCAPE python
+            script could not be found.
+        RuntimeError: If BiG-SCAPE fails to run.
+    """
```
```suggestion
    Examples:
        >>>  from nplinker.genomics.bigscape import run_bigscape
        >>> run_bigscape(antismash_path="./antismash", output_path="./output",  
        ... extra_params="--help", version=1)
    """
```

CunliangGeng reviewed (approved):
Last two comments about docstring.

🎉 then you can directly merge the PR!

adraismawur committed:
Apply suggestions from code review

Co-authored-by: Cunliang Geng <c.geng@esciencecenter.nl>

adraismawur committed:
fix ruff complaints
